Mapping ocaml unix APIs to nodejs calls

phated opened the pull request:
This is just a draft right now because I need a lot of advice on TODOs and testing (sorry, first time contributing here).

@hhugo saw my work on these utilities in the Grain repository and said I should look into upstreaming them. So here is my first pass.

I tried changing some of our assumptions to allow for non-nodejs usage, but I don't know what the functions should do if not inside nodejs.

Please help guide me into taking this over the finish line! 🍻

phated commented:
If I'm looking at this correctly, it seems the only way to write tests that run in nodejs is to include them inside the `compiler/tests-compiler/` directory and use `Util.compile_and_run` - is this correct?

phated commented:
@hhugo if you have some time to look over this, I'm hoping to get this over the finish line within the next week-or-so!

TyOverby commented:
> If I'm looking at this correctly, it seems the only way to write tests that run in nodejs is to include them inside the `compiler/tests-compiler/` directory and use `Util.compile_and_run` - is this correct?

Yes, this is correct.  I’m not sure if any of the runtime js files are linked by default though, you may need to make changes in Util

phated commented:
> Yes, this is correct. I’m not sure if any of the runtime js files are linked by default though, you may need to make changes in Util

Hmm, I'm not sure what you mean

hhugo commented on runtime/unix.js:
```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    // TODO: Make sure this works `EUNKNOWNERR(int)`
```
This comment needs to be addressed

phated replied:
This was an old comment, I'm 99% sure this is how the variant would be constructed. I don't know how to actually force this type of error in nodejs, and I don't think we can mock it. I removed the comment in my refactor.

hhugo commented on runtime/unix.js:
```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+  if (js_stats.isFile()) {
+    file_kind = 0;
+  }
+  if (js_stats.isDirectory()) {
```
Are we sure all those ifs are independent ? I would probably use if-else-if ..

phated replied:
Done in my refactor. I believe they are all independent, but we can be careful.

hhugo commented on runtime/unix.js:
```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+//Requires: fs_node_supported, caml_stats_from_js
+//Requires: caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_stat(filename) {
```
Theses functions bypass the filesystem proxy of jsoo. I wonder if this is the right. Let me think about this a bit more

phated replied:
Refactored!

hhugo commented on runtime/unix.js:
```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+//Provides: unix_mkdir
+//Requires: fs_node_supported, caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_mkdir(dirname, perm) {
```
Related to #1109

hhugo replied:
I've implemented Sys.mkdir in #1134, `unix_mkdir` could directly call into caml_sys_mkdir

phated replied:
I investigated this some more and I found that `caml_sys_mkdir` can rethrow a `caml_raise_sys_error`, but we can't reconstruct a `Unix.Unix_error` if done the other way around. So I've refactored a bit to raise the Unix_error and then catch and re-raise in the `caml_sys_*` functions.

phated replied:
I've separated that refactor into it's own commits

hhugo commented:
> > Yes, this is correct. I’m not sure if any of the runtime js files are linked by default though, you may need to make changes in Util
> 
> Hmm, I'm not sure what you mean

That mean `Util.compile_and_run` might not give enough flags when calling the  js_of_ocam compiler,  one might need to tune `util.ml` to do so.

I don't think you need to worry about this as unix.js is already part of the default runtime.

hhugo commented:
> If I'm looking at this correctly, it seems the only way to write tests that run in nodejs is to include them inside the `compiler/tests-compiler/` directory and use `Util.compile_and_run` - is this correct?

This is not the only way. Look at tests in lib/test (e.g. https://github.com/ocsigen/js_of_ocaml/blob/master/lib/tests/test_nodejs_filesystem_errors.ml). Theses tests run in node only, the dune file says ` (inline_tests
    (modes js))`

hhugo commented:
I though more about this. I would prefer to integrate with the "js_of_ocaml filesystem proxy" with something similar to what we have with `caml_sys_rename`
```js
function unix_stat(n){
  var root = resolve_fs_device(n);
  if(!root.device.stat)
    caml_failwith("unix_stat: no implemented");
  root.device.stat(root.rest);
}
```
You would then have update `fs_node.js` with new methods

hhugo commented:
Improves https://github.com/ocsigen/js_of_ocaml/issues/483

phated commented:
> I though more about this. I would prefer to integrate with the "js_of_ocaml filesystem proxy"

Excellent! Happy to do this. You are okay with JS built for the browser failing with `"unix_stat: not implemented"`?

hhugo commented:
> > I though more about this. I would prefer to integrate with the "js_of_ocaml filesystem proxy"
> 
> Excellent! Happy to do this. You are okay with JS built for the browser failing with `"unix_stat: not implemented"`?

That or we implement some dummy stat.

phated commented:
> That or we implement some dummy stat.

One thing I considered for Grain was embedding the file stats of the built JS file for anything in memory, since any embedded files would be a snapshot at that timestamp. Is this a good idea? If so, any thoughts on how I'd implement that?

hhugo commented:
Can you describe the grain use case a bit more ?

phated commented:
> Can you describe the grain use case a bit more ?

Currently, Grain compiles the compiler (written in Reason) to JS using JSOO and then includes it into a nodejs runtime called [`pkg`](https://www.npmjs.com/package/pkg). `pkg` provides the filesystem abstraction for us as long as the resulting JS calls nodejs APIs that we expect (the current implementation of this logic). Eventually the primary compiler will move to native executables.

However, I also did some research into a browser bundle to add a playground on our website—this should be a natural workflow, as we can compile to WebAssembly in the browser, and then directly run it in a WebAssembly context. Unfortunately, we'd need the unix methods supported in the browser because our compiler compares timestamps of files for caching and recompilation. I was planning to add this feature in the future when work began on our in-browser editor.

hhugo commented:
The current implementation of fs_fake.js is a bit naive currently. With some work you should be able to add  additional information like stat.

hhugo commented:
I don't have more to say right now, but I'm happy to answer any question if you spend some time working on it.

phated commented:
Thanks! For this first draft, I'll not handle the browser APIs, but I will plan to revisit in the future.

phated commented on runtime/fs_node.js:
```diff
@@ -40,30 +41,31 @@ MlNodeDevice.prototype.exists = function(name) {
     return this.fs.existsSync(this.nm(name))?1:0;
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    // TODO: Is it correct to always return a boolean, even on error?
+    return 0;
```
@hhugo should this **ever** raise or would it just always return a boolean?

hhugo replied:
Always returning a boolean seems fine. I don't think the native ocaml runtime would ever raise.

phated commented:
@hhugo this should be ready for review! I added some tests that I could think of for the Unix calls that were implemented. Let me know if you think there should be more tests.

hhugo commented on runtime/fs.js:
```diff
@@ -189,24 +189,32 @@ function caml_sys_file_exists (name) {
+    if(ok == 0) caml_raise_no_such_file(name);
+    return 0;
+  } catch (err) {
+    caml_raise_sys_error(name.toString());
```
actually, fs_node already catch errors from unlink and convert it to sys_error

hhugo commented on runtime/fs.js:
```diff
@@ -189,24 +189,32 @@ function caml_sys_file_exists (name) {
-    l[i+1] = caml_string_of_jsbytes(a[i]);
-  return l;
+  try {
+    var a = root.device.readdir(root.rest);
```
In fact, the try catch is already done in fs_node.js

hhugo commented on runtime/fs.js:
```diff
@@ -218,15 +226,19 @@ function caml_sys_is_directory(name){
   if(!o_root.device.rename)
     caml_failwith("caml_sys_rename: no implemented");
-  o_root.device.rename(o_root.rest, n_root.rest);
+  try {
```
try-catch already done in fs_node.js

hhugo commented on runtime/fs_node.js:
```diff
@@ -40,30 +41,31 @@ MlNodeDevice.prototype.exists = function(name) {
     return 0
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    this.raise_unix_exn_of_nodejs_error(err);
```
I've another proposal, have mkdir take a third argument and decide what error to raise based on that. What do you think ?

phated replied:
I think that will work. My issue was trying to raise the appropriate exception type depending on where this was called from. A flag would work better than the extra try-catch logic I added.

hhugo commented on runtime/fs_node.js:
```diff
@@ -40,30 +41,31 @@ MlNodeDevice.prototype.exists = function(name) {
     return 0
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    this.raise_unix_exn_of_nodejs_error(err);
```
same here

hhugo commented on runtime/fail.js:
```diff
@@ -26,6 +26,9 @@ function caml_return_exn_constant (tag) { return tag; }
 function caml_raise_with_arg (tag, arg) { throw [0, tag, arg]; }
 
+//Provides: caml_raise_with_args (const, const, const)
+function caml_raise_with_args (tag, nargs, args) { throw [0, tag].concat(args); }
```
Can we validate, maybe truncate `args` based on `nargs` ? Alternatively, we could completely remove `nargs`, this function is only a helper for writing stubs

phated replied:
Oh, so the function that exists in ocaml core isn't used anywhere except for external stubs? Would JSOO want to provide the same interface that it provides? I'm happy to add a validation or truncate.

hhugo replied:
I don't think we need to keep the same interface in the case. I would vote for removing the unused argument

hhugo reviewed (commented):
It took me a while get the full picture of changes. Some comments probably reflect that.
The one thing I don't like about it is the raise-catch-raise business. see my comment about adding an extra argument.

hhugo commented on runtime/fs_node.js:
```diff
@@ -104,16 +106,136 @@ MlNodeDevice.prototype.open = function(name, f) {
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    variant = BLOCK(0, err.errno);
```
A comment explaining this is EUNKNOWNERR would be nice. What if errno is not defined ?

phated replied:
I believe errno is always supposed to be attached to a node "fs" error. Do you want me to add a fallback int?

phated replied:
Comment added, let me know if there's a fallback in that makes sense 👍

phated commented on runtime/fs.js:
```diff
@@ -189,24 +189,32 @@ function caml_sys_file_exists (name) {
+      l[i+1] = caml_string_of_jsbytes(a[i]);
+    return l;
+  } catch (err) {
+    caml_raise_sys_error(name.toString());
```
I'm fine with switching back to `err.toString()` for Sys_error exceptions (in combination with the flag mentioned below to raise Unix_errors); however, I reviewed the ocaml stdlib and it often raises a Sys_error with only the filepath. I'm guessing JSOO doesn't use only the filepath for better runtime debugging?

hhugo replied:
It seems the ocaml runtime often calls `caml_sys_error(filename)` that in turn tried to construct an informative message before raising `Sys_error`

phated commented:
@hhugo Thanks for the feedback! I've updated the MlNodeDevice methods to take the `raise_unix` flag as the last argument and use that to determine if Sys_error or Unix_error exceptions are raised.

There are a couple decisions to be made still (unresolved comments) but I'm happy to get those fixed up in the morning.

hhugo commented on runtime/fs.js:
```diff
@@ -189,7 +189,7 @@ function caml_sys_file_exists (name) {
 //Provides: caml_sys_read_directory
 //Requires: caml_string_of_jsbytes
-//Requires: caml_raise_not_a_dir, resolve_fs_device
+//Requires: caml_raise_not_a_dir, resolve_fs_device, caml_raise_sys_error
```
try to keep the `//Requires:` accurate

phated replied:
Thanks! Updated.

hhugo commented on runtime/fs_node.js:
```diff
@@ -104,16 +121,161 @@ MlNodeDevice.prototype.open = function(name, f) {
+    caml_string_of_jsbytes(err.syscall || ""),
+    caml_string_of_jsbytes(err.path || "")
+  ];
+  caml_raise_with_args(caml_named_value("Unix.Unix_error"), args.length, args);
```
We might want to have a fallback in case `Unix.Unix_error` was not registered.

phated replied:
Is there a way to determine if `Unix.Unix_error` was registered inside the JS?

hhugo replied:
`if(caml_named_value("Unix.Unix_error"))`

phated replied:
Is it fine to fallback to the Sys_error if `Unix.Unix_error` is not registered? I changed the `if` statement to `if (raise_unix && caml_named_value("Unix.Unix_error")) {`

hhugo replied:
looks good. Maybe make a single call to caml_named_value.

hhugo commented on runtime/fs_node.js:
```diff
@@ -104,16 +121,161 @@ MlNodeDevice.prototype.open = function(name, f) {
     this.fs.renameSync(this.nm(o), this.nm(n));
   } catch (err) {
-    caml_raise_sys_error(err.toString());
+    if (raise_unix) {
```
I would refactor some of the error handling logic and lift things to a helper function.
```js
function raise_nodejs_error(err, raise_unix) {
    if (raise_unix) {
      this.raise_unix_exn_of_nodejs_error(err);
    } else {
      caml_raise_sys_error(err.toString());
    }
}
```

phated replied:
Updated! I'm going to refactor this a bit more so the nodejs error transformation just happens in this helper.

hhugo commented on runtime/fs_node.js:
```diff
@@ -104,17 +126,125 @@ MlNodeDevice.prototype.open = function(name, f) {
+MlNodeDevice.prototype.readlink = function(name, raise_unix) {
+  try {
+    var js_stats = this.fs.readlinkSync(this.nm(name), 'utf8');
+    return caml_string_of_jsbytes(js_stats);
```
- please rename js_stats.
- do we want to call caml_string_of_jsbytes or caml_string_of_jsstring ?

phated replied:
I've renamed the variable. I'm actually still unclear about the difference between `caml_string_of_jsbytes` and `caml_string_of_jsstring`? If I change this to `caml_string_of_jsstring`, do I also need to change the 2 references inside `raise_nodejs_error`?

phated replied:
I'm noticing that `caml_string_of_jsstring` has the `//If: js-string` requirement. What are the implications of that if this uses `caml_string_of_jsstring`?

We aren't using the `js-string` option in Grain so I'm unfamiliar.

hhugo replied:
```//If: js-string``` is to have conditional stubs based on the js-string option see https://github.com/ocsigen/js_of_ocaml/pull/923. In short, one can choose the memory representation of ocaml strings.

`caml_string_of_jsbyte` should be used when the javascript string represent sequence of bytes, where all chars fall into the 0..255 range. `caml_string_of_jsstring` should be used for normal javascript strings, utf16 encoded. The utf8 encoding  argument is confusing here but I think you still get back a regular javascript string, utf16 encoded. I would use caml_string_of_jsstring and update the two other references.

hhugo commented:
It almost feels you're sitting next to me :)

hhugo commented on compiler/tests-compiler/unix_fs.ml:
```diff
@@ -19,8 +19,14 @@
 
 open Util
 
+(* Unix.LargeFile requires unix_stat_64 which causes tests
```
I don't understand why this is only an issue on windows.
What about we just alias the primitive to unix_stat ?

phated replied:
I was also confused by that. The type signatures look the same, so it would probably work. I should probably provide the `unix_lstat_64` too: https://github.com/ocaml/ocaml/blob/f68acd1a618ac54790a8347fad466084f15a9a9e/otherlibs/win32unix/unix.ml#L407-L408

hhugo commented on compiler/tests-compiler/sys_fs.ml:
```diff
@@ -50,3 +50,23 @@ let%expect_test _ =
+    {|
+  (match Sys.mkdir "/not/exists" 0o777 with
+  | exception Sys_error path -> print_endline ("Sys_error: " ^ path)
+  | exception _ -> print_endline "INCORRECT ERROR"
```
Printing the exception will be more informative in the error case.

phated replied:
I was just trying to debug some Windows test errors and thought the same thing 😅 Updated now.

phated commented:
The MacOS runs have been really flakey on my fork. It seems to be something with the opam env?

I believe there's still 1 windows test failure I need to handle, but my run should finish soon.

hhugo commented:
> The MacOS runs have been really flakey on my fork. It seems to be something with the opam env?
> 
> I believe there's still 1 windows test failure I need to handle, but my run should finish soon.

@smorimoto, the macos CI look rather unstable, I don't know if there is anything we can do about it ?

phated commented:
I believe with that last push that the tests should run on Windows. There were a couple issues:
1. The JSOO program on Windows raises a Sys_error when `Unix.mkdir` or `Unix.rmdir` use an absolute path, like `"/not/exists"`
2. The Unix_errors raised by Windows are often different than their POSIX counterparts. I couldn't find a good way to handle that with ppx_expect so I just made the tests handle both cases where necessary.
3. Tests don't log filepaths anymore since those aren't portable across operating systems.

phated commented on compiler/tests-compiler/unix_fs.ml:
```diff
@@ -172,11 +175,14 @@ let%expect_test "Unix.readlink_EINVAL" =
+  (* Linux & Mac raise Unix.EINVAL *)
+  | exception Unix.Unix_error (Unix.EINVAL, syscall, path) -> print_endline "EXPECTED ERROR"
+  (* Windows raises Unix.EUNKNOWNERR *)
+  | exception Unix.Unix_error (Unix.EUNKNOWNERR(code), syscall, path) -> print_endline "EXPECTED ERROR"
```
We also have a `Unix.EUNKNOWNERR` now because Windows doesn't know how to handle readlink on a directory.

hhugo commented:
> I believe with that last push that the tests should run on Windows. There were a couple issues:
> 
> 1. The JSOO program on Windows raises a Sys_error when `Unix.mkdir` or `Unix.rmdir` use an absolute path, like `"/not/exists"`
> 2. The Unix_errors raised by Windows are often different than their POSIX counterparts. I couldn't find a good way to handle that with ppx_expect so I just made the tests handle both cases where necessary.
> 3. Tests don't log filepaths anymore since those aren't portable across operating systems.

For 1., Maybe we should fix the tests to use a valid root based on the system.

hhugo commented:
Were you able to test grain with this PR ? Do you still this "hack"/jsoo_headers/runtime changes after that ? (Except maybe for always mounting c:/)

smorimoto commented:
The dune cache does make things pretty bad in some cases. It's fixed in dune 3, but the problem can still happen in 2. So I was thinking about backporting cache functionality from 3 to 2, but I hadn't even started discussing that yet.

phated commented:
> For 1., Maybe we should fix the tests to use a valid root based on the system.

I don't think it's entirely necessary to test the absolute paths, but if you think it makes sense, I can look into how that would be done. I think it would be hard because JSOO always tells the program it is in a unix env.

> Were you able to test grain with this PR ? Do you still this "hack"/jsoo_headers/runtime changes after that ? (Except maybe for always mounting c:/)

I'll be testing this today. I believe the only thing Grain will need is to always mount the C: drive due to our packaging mechanisms.

hhugo commented:
> I don't think it's entirely necessary to test the absolute paths, but if you think it makes sense, I can look into how that would be done.

Using unix path on windows make us test `resolve_fs_device` only.
If relative path allow us to test correctly, I'm fine with it.
 
> I think it would be hard because JSOO always tells the program it is in a unix env.

This is no longer true, as of #1132

hhugo commented on compiler/tests-compiler/unix_fs.ml:
```diff
@@ -0,0 +1,217 @@
+  let l = Sys.readdir "aaa" |> Array.to_list in
+  List.iter print_endline l;
+  (match Unix.rmdir "aaa" with
+  | exception _ -> ()
```
Could we restrict to exception we expect

hhugo replied:
I've done that

hhugo commented on compiler/tests-compiler/unix_fs.ml:
```diff
@@ -0,0 +1,217 @@
+  let l = Sys.readdir "aaa" |> Array.to_list in
+  List.iter print_endline l;
+  (match Unix.rmdir "aaa" with
+  | exception _ -> ()
```
Same here

hhugo replied:
I've done that

phated commented:
> This is no longer true, as of #1132

In that PR, I believe you fixed the Windows pathing issue, but I think `Sys.win32` still always returns false. How should I build the absolute path independent of the filesystem?

hhugo commented:
One more thing, could you add tests for the new functions but under the `/static/`directory. The tests should show that some features are not implemented yet. It will be useful if you/one decide to add the support.

hhugo commented:
> > This is no longer true, as of #1132
> 
> In that PR, I believe you fixed the Windows pathing issue, but I think `Sys.win32` still always returns false. How should I build the absolute path independent of the filesystem?

I think Sys.os_type should return Cygwin on windows. But indeed, I haven't checked other values.

hhugo commented:
I've changed the tests related to mkdir and rmdir to also run on the '/static' root and fixed fs_fake accordingly.

hhugo commented on runtime/fs_fake.js:
```diff
@@ -62,15 +63,64 @@ MlFakeDevice.prototype.exists = function(name) {
+  var unix_error = raise_unix && caml_named_value('Unix.Unix_error');
+  if(this.exists(name)) {
+    if (unix_error) {
+      caml_raise_with_args(unix_error, [8, caml_string_of_jsstring("mkdir"), caml_string_of_jsstring(this.nm(name))]);
```
We need a readable way to reference the error constructor.

phated replied:
@hhugo I added another function in https://github.com/ocsigen/js_of_ocaml/pull/1115/commits/fad310827a9b0829ae60572f5281b1f11276c3ff called `make_unix_err_args` that allows the string named to be used. Do you think that is a good solution?

hhugo commented on runtime/fs_fake.js:
```diff
@@ -62,15 +63,64 @@ MlFakeDevice.prototype.exists = function(name) {
   var name_slash = (name == "")?"":(this.slash(name));
   var r = new RegExp("^" + name_slash + "([^/]+)");
+  if(!this.exists(name)) {
+    if (unix_error) {
```
This error handling here is very repetitive, can we do something about it ?

phated replied:
At first glance, I don't see a good way to reduce the repetition. Any helper methods we add for this will require many arguments to handle the variation between different syscalls ("mkdir" and "rmdir" currently). We could implement them to take JS objects like `{ message, code, syscall, path }` and make a helper similar to the `raise_nodejs_error` helper in fs_node.

I'll keep thinking about this.

phated commented:
Sorry for disappearing for a little bit, I was putting together a patch for Grain to ensure this all worked when combined. I drafted a PR on Grain with this code applied at https://github.com/grain-lang/grain/pull/922

hhugo commented:
I've fixed the CI on master. Can you rebase your PR.

phated commented:
> Can you rebase your PR.

Done!

hhugo commented on .github/workflows/build.yml:
```diff
@@ -69,6 +69,9 @@ jobs:
       - run: opam exec -- git diff --exit-code
         if: ${{ matrix.ocaml-compiler == '4.12.0' }}
 
+      - run: opam exec -- git diff --exit-code
```
This is duplicated. You should remove this diff

phated replied:
Good catch. Updated!

hhugo commented:
> One more thing, could you add tests for the new functions but under the `/static/`directory. The tests should show that some features are not implemented yet. It will be useful if you/one decide to add the support.

I few more tests and we should be ready to merge

phated commented:
> few more tests and we should be ready to merge

I'll try to get to that this afternoon. Have a couple of job interviews, but hope to have time outside of them.

phated commented:
I added the tests in the same way you added Unix.mkdir and Unix.rmdir tests, but I added the Failure messages to the expected output. If there is a better way to handle these, please let me know 🙏

hhugo commented:
Thanks a lot for the hard work, I will merge once the CI is happy.
